app/code/Magento/Catalog/Setup/AttributeConfiguration/CatalogConfiguration.php

```diff
@@ -0,0 +1,239 @@
+    {
+        return $this->getNewInstanceWithProperty('visible_in_advanced_search', (bool) $flag);
+    }
+
+    /**
+     * @param bool $flag
+     * @return CatalogConfiguration
+     */
+    public function filterableInSearch($flag = true)
+    {
+        return $this->getNewInstanceWithProperty('filterable_in_search', (bool) $flag);
+    }
+
+    /**
+     * @param bool $flag
+     * @return CatalogConfiguration
+     */
+    public function usedInProductListing($flag = true)
+    {
+        return $this->getNewInstanceWithProperty('used_in_product_listing', (bool) $flag);
+    }
+
+    /**
+     * @param bool $flag
+     * @return CatalogConfiguration
+     */
+    public function usedForSortBy($flag = true)
+    {
+        return $this->getNewInstanceWithProperty('used_for_sort_by', (bool) $flag);
+    }
+
+    /**
+     * @param string[] $productTypes Examples: simple, virtual, bundle, downloadable, configurable
+     * @return CatalogConfiguration
+     */
+    public function applyingTo(array $productTypes)
+    {
+        return $this->getNewInstanceWithProperty('apply_to', implode(',', $productTypes), true);
+    }
+
+    /**
+     * @param int $position
+     * @return CatalogConfiguration
+     * @throws LocalizedException On non-integer position
+     */
+    public function withPosition($position)
+    {
+        if (!is_int($position)) {
+            throw new LocalizedException(__('Non-integer catalog attribute position provided.'));
+        }
+        return $this->getNewInstanceWithProperty('position', $position);
+    }
+
+    /**
+     * @param bool $flag
+     * @return CatalogConfiguration
+     */
+    public function usedForPromoRules($flag = true)
+    {
+        return $this->getNewInstanceWithProperty('used_for_promo_rules', (bool) $flag);
+    }
+
+    /**
+     * @param bool $flag
+     * @return CatalogConfiguration
+     */
+    public function usedInGrid($flag = true)
+    {
+        return $this->getNewInstanceWithProperty('is_used_in_grid', (bool) $flag);
+    }
+
+    /**
+     * @param bool $flag
+     * @return CatalogConfiguration
+     */
+    public function visibleInGrid($flag = true)
+    {
+        return $this->getNewInstanceWithProperty('is_visible_in_grid', (bool) $flag);
+    }
+
+    /**
+     * @param bool $flag
+     * @return CatalogConfiguration
+     */
+    public function filterableInGrid($flag = true)
+    {
+        return $this->getNewInstanceWithProperty('is_filterable_in_grid', (bool) $flag);
+    }
+
+    /**
+     * @return array
+     */
+    public function toArray()
+    {
+        return array_merge(
+            $this->mainConfiguration->toArray(),
+            $this->attributeConfig->toArray()
+        );
+    }
+
+    /**
+     * @return void
+     */
+    public function __clone()
+    {
+        $this->attributeConfig = new DataObject($this->attributeConfig->toArray());
+        $this->mainConfiguration = clone $this->mainConfiguration;
```

Comment:
No reason to clone main configuration here as it's immutable so safely may be shared by several instances